aten/src/ATen/native/cpu/UpSampleKernel.cpp

```diff
@@ -0,0 +1,299 @@
+
+template <typename T>
+inline T data_index_init(T offset) {
+  return offset;
+}
+
+template <typename T, typename... Args>
+inline T data_index_init(T offset, T &x, const T &X, Args &&... args) {
+  offset = data_index_init(offset, std::forward<Args>(args)...);
+  x = offset % X;
+  return offset / X;
+}
+
+inline bool data_index_step() {
+  return true;
+}
+
+template <typename T, typename... Args>
+inline bool data_index_step(T &x, const T &X, Args &&... args) {
+  if (data_index_step(std::forward<Args>(args)...)) {
+    x = ((x + 1) == X) ? 0 : (x + 1);
+    return x == 0;
+  }
+  return false;
+}
+
+static inline int64_t nearest_idx(
+    int64_t output_index,
+    int64_t input_size,
+    int64_t output_size,
+    c10::optional<double> scales) {
+  if (output_size == input_size) {
+    // scale_factor = 1, simply copy
+    return output_index;
+  } else if (output_size == 2 * input_size) {
+    // scale_factor = 2, shift input index
+    return output_index >> 1;
+  } else {
+    float scale = compute_scales_value<float>(scales, input_size, output_size);
+    return nearest_neighbor_compute_source_index(scale, output_index, input_size);
+  }
+}
+
+template <typename scalar_t, typename scale_type>
+void cpu_upsample_nearest(
+    Tensor& output_,
+    const Tensor& input_,
+    const scale_type& scales) {
+  TORCH_CHECK(input_.dtype() == output_.dtype(), "expected dtype ", input_.dtype(),
+              " for `output` but got dtype ", output_.dtype());
+  auto input = input_.contiguous();
+  auto output = output_.contiguous();
+
+  auto input_data = input.data_ptr<scalar_t>();
+  auto output_data = output.data_ptr<scalar_t>();
+  auto input_sizes = input.sizes().vec();
+  auto output_sizes = output.sizes().vec();
+  auto ndim = input_sizes.size();
+  auto numel = output.numel();
+
+  // treat nbatch and channels as one dimension
+  int64_t channels = input_sizes[0] * input_sizes[1];
+  int64_t input_depth = (ndim == 5) ? input_sizes[2] : 1;
+  int64_t output_depth = (ndim == 5) ? output_sizes[2] : 1;
+  int64_t input_height = (ndim >= 4) ? input_sizes[ndim - 2] : 1;
+  int64_t output_height = (ndim >= 4) ? output_sizes[ndim - 2] : 1;
+  int64_t input_width = input_sizes[ndim - 1];
+  int64_t output_width = output_sizes[ndim - 1];
+
+  auto loop1d = [&](int64_t start, int64_t end) {
+    int64_t c = 0;
+    int64_t ow = 0;
+    data_index_init(start, c, channels, ow, output_width);
+    for (int64_t i = start; i < end; i++) {
+      int64_t iw = nearest_idx(ow, input_width, output_width, scales[0]);
+      output_data[i] = input_data[c * input_width + iw];
+      data_index_step(c, channels, ow, output_width);
+    }
+  };
+
+  auto loop2d = [&](int64_t start, int64_t end) {
+    int64_t c = 0;
+    int64_t oh = 0;
+    int64_t ow = 0;
+    data_index_init(start, c, channels, oh, output_height, ow, output_width);
+
+    for (int64_t i = start; i < end; i++) {
+      int64_t ih = nearest_idx(oh, input_height, output_height, scales[0]);
```

Comment:
nit: You can get better performance by creating `output_height`-sized vector and replacing computation with one lookup.

Reply:
yes, right. I checked with vtune this part is already very fast so we don't have to do anything here. This idea should be very valuable for upsample_linear1d/2d/3d, will give it a try in the following patch of optimizing upsample linear.